readme.md
# Heartbeat

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Total Downloads][ico-downloads]][link-downloads]

A simple way to detect when your queue workers are hanging or can't keep up.

## Installation

Via Composer

``` bash
$ composer require mortenscheel/laravel-heartbeat
```

## Configuration

Publish config
``` bash
$ php artisan vendor:publish --tag=heartbeat-config
```

Default configuration
```php
<?php
return [
    /*
     * When enabled, Laravel's sceduler will run the heartbeat:monitor
     * command every minute, which will:
     * 1. Dispatch heartbeat jobs on the configured queues.
     * 2. Send an email to the configured recipients if a
     *    queue exceeds the maximum time between heartbeats.
     */
    'enabled'               => env('HEARTBEAT_ENABLED', false),
    /*
     * Maximum seconds between heartbeats, unless overridden.
     */
    'default_threshold'     => 120,
    /*
     * Which queues to monitor.
     * To set a custom threshold, use the format
     * 'queue_name' => maximum_seconds,
     */
    'queues'                => [
        'default',
    ],
    /*
     * Recipients will receive an "Unhealthy heartbeat warning" e-mail
     * when one or more queues exceed the configured threshold
     */
    'recipients'            => [
    ],
    /*
     * Minimum time between identical e-mail warnings
     */
    'deduplication_minutes' => 1440,
    /*
     * Heartbeats are stored in the application cache (for 24 hours) with this cache prefix
     */
    'cache_prefix'          => 'heartbeat',
];

```

## Usage
Enable heartbeat detection by adding the following variable to `.env`:
```
HEARTBEAT_ENABLED=true
```
If Laravel's Scheduler is running via cron, the `heartbeat:monitor` command will run automatically every minute.

To see the current status of queue heartbeats, you can run `heartbeat:monitor` manually in the console:
```bash
$ php artisan heartbeat:monitor
+---------+----------------+-------------+
| Queue   | Last heartbeat | Max allowed |
+---------+----------------+-------------+
| default | 1 minute ago   | 2 minutes   |
| high    | 1 minute ago   | 2 minutes   |
| low     | 1 minute ago   | 3 minutes   |
| encode  | 2 minutes ago  | 30 minutes  |
+---------+----------------+-------------+
```

## Change log

Please see the [changelog](changelog.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [contributing.md](contributing.md) for details and a todolist.

## Security

If you discover any security related issues, please email morten@mortenscheel.com instead of using the issue tracker.

## Credits

- [Morten Scheel][link-author]
- [All Contributors][link-contributors]

## License

license. Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/mortenscheel/laravel-heartbeat.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/mortenscheel/laravel-heartbeat.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/mortenscheel/laravel-heartbeat
[link-downloads]: https://packagist.org/packages/mortenscheel/laravel-heartbeat
[link-author]: https://github.com/mortenscheel
[link-contributors]: ../../contributors


changelog.md
# Changelog

All notable changes to `Heartbeat` will be documented in this file.

## Version 0.9

Initial release


tests/TestCase.php
<?php

namespace MortenScheel\Heartbeat\Tests;

class TestCase extends \Orchestra\Testbench\TestCase
{
    protected function getPackageProviders($app)
    {
        return ['MortenScheel\Heartbeat\HeartbeatServiceProvider'];
    }
}


tests/HeartbeatDetectionTest.php
<?php

namespace MortenScheel\Heartbeat\Tests;

use MortenScheel\Heartbeat\Heartbeat;
use MortenScheel\Heartbeat\HeartbeatMonitor;

class HeartbeatDetectionTest extends TestCase
{
    /** @var HeartbeatMonitor */
    private $monitor;

    protected function setUp(): void
    {
        parent::setUp();
        $this->monitor = $this->app->make(HeartbeatMonitor::class);
    }


    public function test_only_default_queue_is_monitored()
    {
        $queues = $this->monitor->normalizeQueues()->keys();
        $this->assertContains('default', $queues);
        $this->assertCount(1, $queues);
    }

    public function test_no_heartbeat_are_recorded()
    {
        $heartbeats = $this->monitor->getLatestHeartbeats();
        $this->assertCount(1, $heartbeats);
        $this->assertFalse($heartbeats->first()->wasDetected());
    }

    public function test_heartbeat_can_be_recorded()
    {
        $this->monitor->register('default', 1);
        $heartbeats = $this->monitor->getLatestHeartbeats();
        $this->assertTrue($heartbeats->first()->wasDetected());
    }

    public function test_heartbeats_are_identified_as_unhealthy()
    {
        $this->monitor->register('default', 60);
        $this->assertTrue($this->monitor->getLatestHeartbeats()->first()->isHealthy());
        $this->travel(2)->minutes();
        $this->assertFalse($this->monitor->getLatestHeartbeats()->first()->isHealthy());
    }

    public function test_multiple_queues_can_be_monitored()
    {
        config()->set('heartbeat.queues', [
            'default',
            'low',
            'high',
            'short-threshold'   => 10,
            'shorter-threshold' => 5
        ]);
        $this->monitor->dispatchHeartbeatJobs();
        $this->travel(30)->seconds();
        $unhealthy = $this->monitor->getLatestHeartbeats()->filter(function (Heartbeat $heartbeat) {
            return !$heartbeat->isHealthy();
        });
        $this->assertCount(2, $unhealthy);
    }
}


src/Heartbeat.php
<?php

namespace MortenScheel\Heartbeat;

use Carbon\Carbon;

class Heartbeat
{
    /** @var int */
    protected $threshold;
    /** @var string */
    private $queue;
    /** @var \Carbon\Carbon|null */
    private $executed_at;

    public function __construct(
        string $queue,
        int $threshold,
        Carbon $executed_at = null
    ) {
        $this->queue = $queue;
        $this->threshold = $threshold;
        $this->executed_at = $executed_at;
    }

    public function wasDetected(): bool
    {
        return $this->executed_at !== null;
    }

    public function isHealthy(): bool
    {
        return $this->wasDetected() &&
            $this->executed_at->diffInSeconds() <= $this->threshold;
    }

    public function diffForHumans(): string
    {
        return $this->wasDetected() ? $this->executed_at->diffForHumans() : 'Not found';
    }

    public function maxTimeForHumans(): string
    {
        return now()->subSeconds($this->threshold)->diffForHumans(null, true);
    }

    public function getQueue(): string
    {
        return $this->queue;
    }

    public function getExecutedAt(): ?Carbon
    {
        return $this->executed_at;
    }
}


src/HeartbeatMonitor.php
<?php

namespace MortenScheel\Heartbeat;

use Illuminate\Support\Collection;
use Illuminate\Support\Facades\Cache;
use Illuminate\Support\Facades\Mail;
use MortenScheel\Heartbeat\Jobs\HeartbeatJob;
use MortenScheel\Heartbeat\Mail\UnhealthyHeartbeatMail;
use function is_string;

class HeartbeatMonitor
{
    public function isEnabled(): bool
    {
        return config('heartbeat.enabled');
    }

    public function register(string $queue, int $threshold): void
    {
        Cache::put(
            $this->getCacheKey($queue),
            new Heartbeat($queue, $threshold, now()),
            now()->addHours(24)
        );
    }

    /**
     * @return \Illuminate\Support\Collection|array<string, int>
     */
    public function normalizeQueues(): Collection
    {
        return collect(config('heartbeat.queues'))->mapWithKeys(function ($value, $key) {
            if (is_string($key)) {
                return [$key => $value];
            }
            return [$value => config('heartbeat.default_threshold')];
        });
    }

    public function dispatchHeartbeatJobs(): void
    {
        $this->normalizeQueues()->each(function (int $threshold, string $queue) {
            HeartbeatJob::dispatch($queue, $threshold);
        });
    }

    /**
     * @return \Illuminate\Support\Collection|\MortenScheel\Heartbeat\Heartbeat[]
     */
    public function getLatestHeartbeats(): Collection
    {
        return $this->normalizeQueues()->map(function (int $threshold, string $queue) {
            return Cache::get($this->getCacheKey($queue)) ?? new Heartbeat($queue, $threshold);
        })->values();
    }

    public function warnRecipients(Collection $unhealthy): void
    {
        $queues = $unhealthy->map->getQueue()->toArray();
        $key = $this->getCacheKey('previous_warning');
        if (($previous_warning = Cache::get($key)) && $previous_warning === $queues) {
            return;
        }
        Mail::to(config('heartbeat.recipients'))->send(new UnhealthyHeartbeatMail($unhealthy));
        Cache::put($key, $queues, now()->addMinutes(config('heartbeat.deduplication_minutes')));
    }

    private function getCacheKey(string $name): string
    {
        return sprintf('%s:%s', config('heartbeat.cache_prefix'), $name);
    }
}


src/Jobs/HeartbeatJob.php
<?php

namespace MortenScheel\Heartbeat\Jobs;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use MortenScheel\Heartbeat\HeartbeatMonitor;

class HeartbeatJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    /** @var int */
    protected $threshold;

    /**
     * Create a new job instance.
     * @return void
     */
    public function __construct(string $queue, int $threshold)
    {
        $this->queue = $queue;
        $this->threshold = $threshold;
    }


    /**
     * Execute the job.
     * @return void
     */
    public function handle()
    {
        (new HeartbeatMonitor)->register($this->queue, $this->threshold);
    }
}


src/HeartbeatServiceProvider.php
<?php

namespace MortenScheel\Heartbeat;

use Illuminate\Console\Scheduling\Schedule;
use Illuminate\Support\ServiceProvider;
use MortenScheel\Heartbeat\Commands\MonitorHeartbeatCommand;

class HeartbeatServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     * @return void
     */
    public function boot(): void
    {
        if ($this->app->runningInConsole()) {
            $this->publishes([
                __DIR__.'/../config/heartbeat.php' => config_path('heartbeat.php'),
            ], 'heartbeat-config');
            $this->publishes([
                __DIR__.'/../resources/views' => resource_path('views/vendor/heartbeat'),
            ], 'heartbeat-views');
            $this->loadViewsFrom(__DIR__.'/../resources/views', 'heartbeat');
            $this->commands([
                MonitorHeartbeatCommand::class
            ]);
            if (config('heartbeat.enabled')) {
                $this->app->booted(function () {
                    $this->app->make(Schedule::class)
                        ->command('heartbeat:monitor')
                        ->everyMinute();
                });
            }
        }
    }

    /**
     * Register any package services.
     * @return void
     */
    public function register(): void
    {
        $this->mergeConfigFrom(__DIR__.'/../config/heartbeat.php', 'heartbeat');
        // Register the service the package provides.
        $this->app->singleton('heartbeat', function ($app) {
            return new HeartbeatMonitor;
        });
    }
}


src/Mail/UnhealthyHeartbeatMail.php
<?php

namespace MortenScheel\Heartbeat\Mail;

use Illuminate\Bus\Queueable;
use Illuminate\Mail\Mailable;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Collection;

class UnhealthyHeartbeatMail extends Mailable
{
    use Queueable, SerializesModels;

    /** @var \Illuminate\Support\Collection */
    protected $heartbeats;

    /**
     * Create a new message instance.
     * @return void
     */
    public function __construct(Collection $heartbeats)
    {
        $this->heartbeats = $heartbeats;
    }

    /**
     * Build the message.
     * @return $this
     */
    public function build()
    {
        return $this->markdown('heartbeat::unhealthy_heartbeat_mail')
            ->with('heartbeats', $this->heartbeats)
            ->subject('Unhealthy heartbeat warning');
    }
}


src/Commands/MonitorHeartbeatCommand.php
<?php

namespace MortenScheel\Heartbeat\Commands;

use Illuminate\Console\Command;
use Illuminate\Support\Collection;
use MortenScheel\Heartbeat\Heartbeat;
use MortenScheel\Heartbeat\HeartbeatMonitor;

class MonitorHeartbeatCommand extends Command
{
    /**
     * The name and signature of the console command.
     * @var string
     */
    protected $signature = 'heartbeat:monitor {--no-dispatch : Do not dispatch heartbeat jobs}
                                              {--no-warn     : Do not send e-mail warnings}';

    /**
     * The console command description.
     * @var string
     */
    protected $description = 'Dispatch and monitor queue heartbeats';

    /**
     * Execute the console command.
     * @return int
     */
    public function handle(HeartbeatMonitor $monitor)
    {
        if (!$monitor->isEnabled()) {
            $this->warn('Heartbeat is not enabled');
            return 1;
        }
        if (!$this->option('no-dispatch')) {
            $monitor->dispatchHeartbeatJobs();
        }
        $heartbeats = $monitor->getLatestHeartbeats();
        $this->renderHeartbeatsTable($heartbeats);
        $unhealthy = $heartbeats->filter(function (Heartbeat $heartbeat) {
            return !$heartbeat->isHealthy();
        });
        if ($unhealthy->isNotEmpty() && !$this->option('no-warn')) {
            $monitor->warnRecipients($unhealthy);
        }
        return 0;
    }

    private function renderHeartbeatsTable(Collection $heartbeats): void
    {
        $headers = ['Queue', 'Last heartbeat', 'Max allowed'];
        $rows = $heartbeats->map(function (Heartbeat $heartbeat) {
            $message = sprintf('<fg=%s>%s</>', ($heartbeat->isHealthy() ? 'green' : 'red'), $heartbeat->diffForHumans());
            return [
                $heartbeat->getQueue(),
                $message,
                $heartbeat->maxTimeForHumans()
            ];
        });
        $this->table($headers, $rows);
    }
}


config/heartbeat.php
<?php
return [
    /*
     * When enabled, Laravel's scheduler will run the heartbeat:monitor
     * command every minute, which will:
     * 1. Dispatch heartbeat jobs on the configured queues.
     * 2. Send an email to the configured recipients if a
     *    queue exceeds the maximum time between heartbeats.
     */
    'enabled'               => env('HEARTBEAT_ENABLED', false),
    /*
     * Maximum seconds between heartbeats, unless overridden.
     */
    'default_threshold'     => 120,
    /*
     * Which queues to monitor.
     * To set a custom threshold, use the format
     * 'queue_name' => maximum_seconds,
     */
    'queues'                => [
        'default',
    ],
    /*
     * Recipients will receive an "Unhealthy heartbeat warning" e-mail
     * when one or more queues exceed the configured threshold
     */
    'recipients'            => [
    ],
    /*
     * Minimum time between identical e-mail warnings
     */
    'deduplication_minutes' => 1440,
    /*
     * Heartbeats are stored in the application cache (for 24 hours) with this cache prefix
     */
    'cache_prefix'          => 'heartbeat',
];
